packages/callhierarchy/src/browser/callhierarchy-service-impl.ts

```diff
@@ -0,0 +1,276 @@
+export class CallHierarchyContext {
+
+    readonly symbolCache = new Map<string, SymbolInformation[]>();
+    readonly disposables: Disposable[] = [];
+
+    constructor(readonly languageClient: ILanguageClient,
+        readonly textModelService: MonacoTextModelService,
+        readonly logger: ILogger) { }
+
+    async getAllSymbols(uri: string): Promise<SymbolInformation[]> {
+        const cachedSymbols = this.symbolCache.get(uri);
+        if (cachedSymbols) {
+            return cachedSymbols;
+        }
+        const symbols = await this.languageClient.sendRequest(DocumentSymbolRequest.type, <DocumentSymbolParams>{
+            textDocument: TextDocumentIdentifier.create(uri)
+        });
+        this.symbolCache.set(uri, symbols);
+        return symbols;
+    }
+
+    async getEditorModelReference(uri: string) {
+        const model = await this.textModelService.createModelReference(new URI(uri));
+        this.disposables.push(model);
+        return model;
+    }
+
+    async getDefinitionLocation(location: Location): Promise<Location | undefined> {
+        const uri = location.uri;
+        const { line, character } = location.range.start;
+        let locations: Location | Location[] | null = null;
+        try {
+            locations = await this.languageClient.sendRequest(DefinitionRequest.type, <TextDocumentPositionParams>{
+                position: Position.create(line, character),
+                textDocument: { uri }
+            });
+        } catch (error) {
+            this.logger.error(`Error from definitions request: ${uri}#${line}/${character}`, error);
+        }
+        if (!locations) {
+            return undefined;
+        }
+        return Array.isArray(locations) ? locations[0] : locations;
+    }
+
+    async getCallerReferences(definition: Location): Promise<Location[]> {
+        try {
+            const references = await this.languageClient.sendRequest(ReferencesRequest.type, <ReferenceParams>{
+                context: {
+                    includeDeclaration: false // TODO find out, why definitions are still contained
+                },
+                position: {
+                    line: definition.range.start.line,
+                    character: definition.range.start.character
+                },
+                textDocument: {
+                    uri: definition.uri
+                }
+            });
+            const uniqueReferences = utils.filterUnique(references);
+            const filteredReferences = utils.filterSame(uniqueReferences, definition);
+            return filteredReferences;
+        } catch (error) {
+            this.logger.error(`Error from references request`, error);
+            return [];
+        }
+    }
+
+    dispose() {
+        this.disposables.forEach(element => {
+            element.dispose();
+        });
+        this.symbolCache.clear();
+    }
+}
+
+@injectable()
+export abstract class CallHierarchyServiceImpl implements CallHierarchyService {
```

Comment:
Would be good to give it a name that indicates its best-effort implementation. Also `Abstract` prefix would be good.

Reply:
Changed to `AbstractDefaultCallhierarchyService`